### PR TITLE
chore(splunk hec sink)!: json encoding should get message first and then parse

### PR DESCRIPTION
**Pull request by pront**

### CODE REVIEW

**Comment by pront** [comment]
Putting this out there to potentially get some feedback, I will revisit on Monday.

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (pront) [inline src/sinks/splunk_hec/logs/encoder.rs:100]: Encoders call `get_message()` internally.

**Review by sebastiantia — APPROVED** [review]

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (pront) [inline src/sinks/splunk_hec/logs/encoder.rs:90]: TODO: This is expected to break all integration tests using `JsonSerializerConfig`. For example, look at [splunk_custom_fields](https://github.com/vectordotdev/vector/blob/v0.37.1/src/sinks/splunk_hec/logs/integration_tests.rs#L355), looking up by `message` is no longer correct. The code needs to be updated to directly access the value from the event.

Comment (pront) [reply src/sinks/splunk_hec/logs/encoder.rs:90]: cc @vladimir-dd and @jszwedko for confirming before I go ahead and edit all those.

**Review by sebastiantia — CHANGES_REQUESTED** [review]
Didn't mean to approve this PR 😓

**Comment by pront** [comment]
> Didn't mean to approve this PR 😓

I think this will be blocked until you re-approve. I suggest re-approving 😅 
I won't merge until Vlad and Jesse take a look.

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (vladimir-dd) [reply src/sinks/splunk_hec/logs/encoder.rs:90]: I think we should move all other fields from LogEvent to HecData.fields otherwise we could lost some fields.

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (vladimir-dd) [inline src/sinks/splunk_hec/logs/encoder.rs:104]: I think we should use the same logic as above here to avoid losing information - but instead of encoding message encode the whole event?

**Review by vladimir-dd — CHANGES_REQUESTED** [review]
Left some comments.

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (vladimir-dd) [reply src/sinks/splunk_hec/logs/encoder.rs:90]: Sorry, didn't notice that we pass fields through `metadata.fields`. But since we do not expose `indexed_fields` for the sink, do we always pass empty`.fields`?

**Review by sebastiantia — APPROVED** [review]

File: src/sinks/splunk_hec/logs/integration_tests.rs
Comment (pront) [inline src/sinks/splunk_hec/logs/integration_tests.rs:527]: 🗒️ `cargo vdev -v int start -a splunk` works in Ubuntu. Also, when working on Ubuntu `cargo vdev -v int test --retries 2 -a splunk` fails to runs tests.

File: changelog.d/splunk_hec_sink_encode_message_only.breaking.md
Comment (jszwedko) [inline changelog.d/splunk_hec_sink_encode_message_only.breaking.md:2]: A few questions:

- Do we know how Splunk reacts if you send it an event with an empty `event` field?
- Is the semantic field required? I.e. will Vector fail to boot if there is no semantic `message` defined?
- Also, is this only true for the `/event` endpoint and not for `/raw`?

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (jszwedko) [reply src/sinks/splunk_hec/logs/encoder.rs:90]: I feel like I'm getting a bit turned around by this now 😵

I think I'd talked to Pavlos about only encoding the message field, but, thinking about it more now, the field in the HEC payload is actually called `event` which makes me think maybe it _should_ be encoding the entire event and not just the `message` key. Their definition of `event`:

> Event data can be assigned to the "event" key within the JSON object in the HTTP request, or it can be raw text. The "event" key is at the same level within the JSON event packet as the metadata keys. Within the "event" key-value curly brackets, the data can be in whatever format you want: a string, a number, another JSON object, and so on.
> 
> You can batch multiple events in one event packet by combining them within the request. By batching the events, you're specifying that any event metadata within the request is to apply to all of the events contained in the request. Batching events can significantly speed performance when you need to index large quantities of data.

https://docs.splunk.com/Documentation/Splunk/9.2.1/Data/FormateventsforHTTPEventCollector#Event_data:~:text=from%20the%20request.-,Event%20data,-Event%20data%20can

`fields` is defined as:

> The fields key isn't applicable to raw data. This key specifies a JSON object that contains a flat (not nested) list of explicit custom fields to be defined at index time. Requests containing the "fields" property must be sent to the /collector/event endpoint, or else they aren't indexed. For more information, see [Indexed field extractions](http://docs.splunk.com/Documentation/Splunk/9.2.1/Data/IFXandHEC).

I agree with @vladimir-dd that if we only encode `message` as `event` that maybe the rest of the fields should go into `fields`; however, now I'm not so sure that the entire event shouldn't just go in `event`.

This might be a case where we need to play around with it a bit and see how it looks like in the Splunk interface under various conditions (like Splunk HEC -> Splunk and Syslog -> Splunk).

Apologies for what feels like going in circles a bit here. I'm happy to chat about this if it helps resolve it more quickly.

**Comment by neuronull** [comment]
👋  
Noting that we elected to hold off on this for the time being.
We can revisit it if needed next week when Pavlos returns.

Thanks for all the input @jszwedko 🙇

File: src/sinks/splunk_hec/logs/encoder.rs
Comment (pront) [reply src/sinks/splunk_hec/logs/encoder.rs:90]: Marking it as a draft for now. It's not urgent to get to the bottom of this given we have other priorities.

**Pull request closed without merging**
